src/main/java/picard/sam/markduplicates/UmiAwareDuplicateSetIterator.java

```diff
@@ -138,32 +140,41 @@ private void process(final DuplicateSet set) {
             String inferredUmi = representativeRead.getStringAttribute(inferredUmiTag);
 
             for (SAMRecord rec : records) {
-                String currentUmi = rec.getStringAttribute(umiTag);
+
+                String currentUmi = UmiUtil.sanitizeUMI(rec, umiTag);
 
                 if (currentUmi != null) {
                     // All UMIs should be the same length, the code presently does not support variable length UMIs
                     // TODO: Add support for variable length UMIs
-                    if (!haveWeSeenFirstRead) {
-                        metrics.MEAN_UMI_LENGTH = currentUmi.length();
-                        haveWeSeenFirstRead = true;
+
+                    //if the UMI contains a N, we don't want to include it in our other metrics
+                    //but still want to keep track of it
+                    if (currentUmi.contains("N")) {
+
+                        metrics.addUmiObservationN();
+
                     } else {
-                        if (metrics.MEAN_UMI_LENGTH != currentUmi.length()) {
-                            throw new PicardException("UMIs of differing lengths were found.");
+
+                        if (!haveWeSeenFirstRead) {
+                            metrics.MEAN_UMI_LENGTH = currentUmi.length();
+                            haveWeSeenFirstRead = true;
+                        } else {
+                            if (metrics.MEAN_UMI_LENGTH != currentUmi.length()) {
+                                throw new PicardException("UMIs of differing lengths were found.");
+                            }
                         }
+                        // Update UMI metrics associated with each record
+                        metrics.OBSERVED_BASE_ERRORS += hammingDistance(currentUmi, inferredUmi);
```

Comment:
Add a comment here that the hammingDistance includes N as a distance of 1.  But comparing N to N is 0 distance.

Reply:
added